daemon/ZilliqaDaemon.cpp

```diff
@@ -113,238 +170,287 @@ vector<pid_t> getProcIdByName(const string& procName, ofstream& log) {
-          return;
-        }
+    po::notify(vm);
+    m_recovery = vm.count("recovery");
 
-        try {
-          const string params = PubKey[pid] + " " + PrivKey[pid] + " " +
-                                Port[pid] + " " + to_string(v) + " " +
-                                Path[pid] + " " + name + " 2>&1";
-          log << "Trying to restart .." << endl;
-          log << "Params: " << params << endl;
-          log << "Output from shell...(if any)" << endl << endl;
-          log << "\" "
-              << execute(string(restart_zilliqa).append(" ").append(params))
-              << " \"" << endl;
-          log << endl;
-        } catch (runtime_error& e) {
-          log << "ERROR!! " << e.what() << endl;
-        }
-        vector<pid_t> tmp = getProcIdByName(name, log);
-
-        for (pid_t i : tmp) {
-          auto it = find(pids[name].begin(), pids[name].end(), i);
-          if (it == pids[name].end()) {
-            died[i] = false;
-            pids[name].push_back(i);
-            log << "Started new process " << name << " with PiD: " << i << endl;
-          }
-        }
-        PrivKey.erase(pid);
-        PubKey.erase(pid);
-        Port.erase(pid);
-        Path.erase(pid);
-        isDS.erase(pid);
-      }
+    if (vm.count("cseed")) {
+      ZilliqaDaemon::LOG(m_log, "Running Daemon for community seed node.");
+      m_cseed = true;
     }
+  } catch (boost::program_options::required_option& e) {
+    ZilliqaDaemon::LOG(m_log, "ERROR: " + string(e.what()));
+    return ERROR_IN_COMMAND_LINE;
+  } catch (boost::program_options::error& e) {
+    ZilliqaDaemon::LOG(m_log, "ERROR: " + string(e.what()));
+    return ERROR_IN_COMMAND_LINE;
   }
+
+  return SUCCESS;
 }
 
-int main() {
-  pid_t pid_parent, sid;
+int main(int argc, const char* argv[]) {
   ofstream log;
-  log.open("daemon-log-local.txt", fstream::out | fstream::trunc);
-
-  pid_parent = fork();
+  log.open(daemon_log.c_str(), fstream::out | fstream::trunc);
+  pid_t pid_parent = fork();
 
   if (pid_parent < 0) {
-    log << "Failed to fork " << endl;
+    ZilliqaDaemon::LOG(log, "Failed to fork.");
     exit(EXIT_FAILURE);
   }
 
   if (pid_parent > 0) {
-    log << "Started daemon successfully" << endl;
+    ZilliqaDaemon::LOG(log, "Started daemon successfully.");
     exit(EXIT_SUCCESS);
   }
 
   umask(0);
 
-  sid = setsid();
-
-  if (sid < 0) {
-    log << "Unable to set sid" << endl;
-    exit(EXIT_FAILURE);
-  }
-
-  if ((chdir("..")) < 0) {
-    log << "Failed to chdir" << endl;
+  if (setsid() < 0) {
+    ZilliqaDaemon::LOG(log, "Unable to set sid.");
     exit(EXIT_FAILURE);
   }
 
   close(STDIN_FILENO);
   close(STDOUT_FILENO);
   close(STDERR_FILENO);
 
-  unordered_map<string, vector<pid_t>> pids;
-  unordered_map<pid_t, bool> died;
+  ZilliqaDaemon daemon(argc, argv, log);
 
-  initialize(pids, died, log);
   while (1) {
-    MonitorProcess(pids, died, log);
+    daemon.MonitorProcess();
```

Comment:
Can we run MonitorProcess for each program in list.? If enable polling for more than 1 processes if needed in future.
```
   for (const auto&p : programName)
   {
        daemon.MonitorProcess(p);
   }
  ```
Also, if we can pass program names as parameter to daemon.